### ISSUE TITLE
Storage resources are insufficient for the write operation.

### ISSUE THREAD

**Issue by sych74** [minio/minio#2600]
## Expected behaviour

create a bucket 
## Actual behaviour

getting error 
time="2016-09-01T17:24:25Z" level=error msg="Unable to create a bucket." cause="Storage resources are insufficient for the write operation.
## Steps to reproduce the behaviour

6 volumes are mounted via NFS 

node88519-minio123.demo.jelastic.com:/data  /export/volume5 nfs soft,intr,udp,nolock,vers=3,rw  0   0
node88519-minio123.demo.jelastic.com:/data  /export/volume6 nfs soft,intr,udp,nolock,vers=3,rw  0   0
node88519-minio123.demo.jelastic.com:/data  /export/volume3 nfs soft,intr,udp,nolock,vers=3,rw  0   0
node88519-minio123.demo.jelastic.com:/data  /export/volume4 nfs soft,intr,udp,nolock,vers=3,rw  0   0
node88519-minio123.demo.jelastic.com:/data  /export/volume1 nfs soft,intr,udp,nolock,vers=3,rw  0   0
node88519-minio123.demo.jelastic.com:/data  /export/volume2 nfs soft,intr,udp,nolock,vers=3,rw  0   0

start minio server in a 6 drives setup.
+ exec app server /export/volume1 /export/volume2 /export/volume3 /export/volume4 /export/volume5 /export/volume6
## Minio version

Version: DEVELOPMENT.GOGET
Release-Tag: DEVELOPMENT.GOGET
Commit-ID: DEVELOPMENT.GOGET
## System information

Debian GNU/Linux 8

**Comment by harshavardhana** [comment]
> 6 volumes are mounted via NFS
> 
> node88519-minio123.demo.jelastic.com:/data /export/volume5 nfs soft,intr,udp,nolock,vers=3,rw 0 0
> node88519-minio123.demo.jelastic.com:/data /export/volume6 nfs soft,intr,udp,nolock,vers=3,rw 0 0
> node88519-minio123.demo.jelastic.com:/data /export/volume3 nfs soft,intr,udp,nolock,vers=3,rw 0 0
> node88519-minio123.demo.jelastic.com:/data /export/volume4 nfs soft,intr,udp,nolock,vers=3,rw 0 0
> node88519-minio123.demo.jelastic.com:/data /export/volume1 nfs soft,intr,udp,nolock,vers=3,rw 0 0
> node88519-minio123.demo.jelastic.com:/data /export/volume2 nfs soft,intr,udp,nolock,vers=3,rw 0 0
> 
> start minio server in a 6 drives setup.
> - exec app server /export/volume1 /export/volume2 /export/volume3 /export/volume4 /export/volume5 /export/volume6

Do you have write access @sych74  on these disks? can you also verify if the backend as following directories created 

```
$ ls -a 
.minio.sys
$ ls -a .minio.sys/
format.json
```

**Comment by harshavardhana** [comment]
> Do you have write access @sych74 on these disks? can you also verify if the backend as following directories created
> 
> $ ls -a 
> .minio.sys
> $ ls -a .minio.sys/
> format.json

@sych74 were you able to verify this again on your end? - Thanks

**Comment by sych74** [comment]
Yes, write access was available, I created couple files manually. Also .minio.sys and format.json were present as well. I recreated a new environment from scratch and it works now, seems like a floating issue. I will play a little bit more with it and will provide feedback how it behaves.

**Comment by harshavardhana** [comment]
> Yes, write access was available, I created couple files manually. Also .minio.sys and format.json were present as well. I recreated a new environment from scratch and it works now, seems like a floating issue. I will play a little bit more with it and will provide feedback how it behaves.

Thanks @sych74

**Comment by harshavardhana** [comment]
@sych74 checking back to see how things are moving - feel free to ping us on gitter or reach out directly @harshavardhana . 

Saw the integration effort here https://github.com/jelastic-jps/minio :+1:

**Comment by sych74** [comment]
So, we have automated installation process to Jelastic, now it's as easy as a click. However, we faced some issues with NFS support. The topology of the environment is the following: 1 minio server container + 6 data containers with NFS servers that are mounted to minio server.  If we shutdown one or two of the data containers we get strange errors like these _input/output error_,  _bucket not found: xxx_. And even more after these containers become available again, minio server shows the same error as it was stated in the issue title _Storage resources are insufficient for the write operation_. 

I can provide an access to our demo lab for a deeper investigation. Please let me know.

**Comment by krishnasrinivas** [comment]
@sych74 

```
node88519-minio123.demo.jelastic.com:/data /export/volume5 nfs soft,intr,udp,nolock,vers=3,rw 0 0
node88519-minio123.demo.jelastic.com:/data /export/volume6 nfs soft,intr,udp,nolock,vers=3,rw 0 0
node88519-minio123.demo.jelastic.com:/data /export/volume3 nfs soft,intr,udp,nolock,vers=3,rw 0 0
node88519-minio123.demo.jelastic.com:/data /export/volume4 nfs soft,intr,udp,nolock,vers=3,rw 0 0
node88519-minio123.demo.jelastic.com:/data /export/volume1 nfs soft,intr,udp,nolock,vers=3,rw 0 0
node88519-minio123.demo.jelastic.com:/data /export/volume2 nfs soft,intr,udp,nolock,vers=3,rw 0 0
```

According to this the minio server is accessing the same nfs backend `node88519-minio123.demo.jelastic.com:/data` for all it's drives, this will be an incorrect configuration.

Is that how it is setup? If `/export/volume<number>` shares the same backend then the erasure file will overwrite each other causing corruption.

**Comment by sych74** [comment]
@krishnasrinivas thanks, good catch, that cut was from first experiments. The latest one is 

```
node88716-env-9340451.demo.jelastic.com:/data   /export/volume0 nfs soft,intr,udp,nolock,retrans=1,retry=0,bg,vers=3,rw 0   0
node88713-env-9340451.demo.jelastic.com:/data   /export/volume1 nfs soft,intr,udp,nolock,retrans=1,retry=0,bg,vers=3,rw 0   0
node88715-env-9340451.demo.jelastic.com:/data   /export/volume2 nfs soft,intr,udp,nolock,retrans=1,retry=0,bg,vers=3,rw 0   0
node88714-env-9340451.demo.jelastic.com:/data   /export/volume3 nfs soft,intr,udp,nolock,retrans=1,retry=0,bg,vers=3,rw 0   0
node88717-env-9340451.demo.jelastic.com:/data   /export/volume4 nfs soft,intr,udp,nolock,retrans=1,retry=0,bg,vers=3,rw 0   0
node88711-env-9340451.demo.jelastic.com:/data   /export/volume5 nfs soft,intr,udp,nolock,retrans=1,retry=0,bg,vers=3,rw 0   0
```

The errors from my previous message appeared with this config as well.

**Comment by sych74** [comment]
There are bit more details and steps to reproduce - create buckets and upload files during the failure simulation 
1. shutdown 1st container - Minio is working fine
2. shutdown 2nd - input/output errors at the beginning, a little bit later seems like no errors
3. shutdown 3nd - storage resources are insufficient for the write operation
4. bringing back 1st - checking write access - ok, but the same error: storage resources are insufficient
5. bringing back 2st - storage resources are insufficient
6. restart minio  (3rd container is still down) - error in logs 
   `+ exec app server --address :80 /export/volume0 /export/volume1 /export/volume2 /export/volume3 /export/volume4 /export/volume5 time="2016-09-07T14:33:55Z" level=fatal msg="Unable to intialize object layer." cause="Inconsistent JBOD found." 
   `

**Comment by harshavardhana** [comment]
> I can provide an access to our demo lab for a deeper investigation. Please let me know.

@sych74  we will take a look, can you arrange access? - Thanks.

**Comment by harshavardhana** [comment]
> There are bit more details and steps to reproduce - create buckets and upload files during the failure simulation
> 
> shutdown 1st container - Minio is working fine
> shutdown 2nd - input/output errors at the beginning, a little bit later seems like no errors
> shutdown 3nd - storage resources are insufficient for the write operation
> bringing back 1st - checking write access - ok, but the same error: storage resources are insufficient
> bringing back 2st - storage resources are insufficient
> restart minio (3rd container is still down) - error in logs + exec app server --address :80 /export/volume0 /export/volume1 /export/volume2 /export/volume3 /export/volume4 /export/volume5 time="2016-09-07T14:33:55Z" level=fatal msg="Unable to intialize object layer." cause="Inconsistent JBOD found."

Do you know the commit-id which you are using here "minio version" can help?  - these errors look like some sort of backend corruption i.e "format.json" 

We would need to see why this is happening,  providing us access would be great. We can also setup a local cluster to test these functionalities - is there easier way?

**Comment by sych74** [comment]
I deployed docker minio/minio:latest, can you provide a hint how to check which commit-id I'm using? 

As regards access, please provide you email I will setup an account for you at the demo lab, and will share the broken env. Also you will be able to create new envs and check how it works independently, the provisioning script is fully automated now, so it should be easy.

**Comment by harshavardhana** [comment]
> As regards access, please provide you email I will setup an account for you at the demo lab, and will share the broken env. Also you will be able to create new envs and check how it works independently, the provisioning script is fully automated now, so it should be easy.

Yeah you can send email to dev < a > minio.io - thanks for your patience.

**Comment by harshavardhana** [comment]
> As regards access, please provide you email I will setup an account for you at the demo lab, and will share the broken env. Also you will be able to create new envs and check how it works independently, the provisioning script is fully automated now, so it should be easy.
> Yeah you can send email to dev < a > minio.io - thanks for your patience.

@sych74 checking back to see if you were able to setup an account on your demo lab ? - Thanks.

**Comment by sych74** [comment]
@harshavardhana, I sent registration and sharing environment requests. Please check it and let me know if see environment Minio (env-9340451.demo.jelastic.com) in your dashboard. After that you can check the logs via UI https://docs.jelastic.com/view-log-files (you need to check run.log) or using ssh access (https://docs.jelastic.com/ssh-access). Please let me know if you are able to perform such actions.

**Comment by harshavardhana** [comment]
> @harshavardhana, I sent registration and sharing environment requests. Please check it and let me know if see environment Minio (env-9340451.demo.jelastic.com) in your dashboard. After that you can check the logs via UI https://docs.jelastic.com/view-log-files (you need to check run.log) or using ssh access (https://docs.jelastic.com/ssh-access). Please let me know if you are able to perform such actions.

@sych74 i have not been able to load "http://app.demo.jelastic.com/" on my end perhaps some network issue.

**Comment by sych74** [comment]
@harshavardhana as I see dev@ account has one environment, is the network issue still present?

**Comment by krishnasrinivas** [comment]
@sych74 we are able to see the dashboard, looking at the issue.

**Comment by krishnasrinivas** [comment]
@sych74 
- we were getting data corruption issues with the setup, but once nfs mount was done using tcp instead of udp, it worked fine (all the data containers were up)
- minio has not been tested with NFS backends.
- when an NFS server is down, operations on the mount point will take long time to return IO error causing the process to stall.
- unfsd is generally not advisable for production deployments.
- We are working on distributed minio XL which will be a good fit for your needs. If you deploy distributed XL you can take out NFS out of the stack and instead deploy minio with this command:

`minio --server server1:/data server2:/data ..... server6:/data`

You have to run the same command on all six servers. you can use docker containers as well - In which case there are no data container and separate minio container like how you have setup. All the 6 containers will be minio containers and each minio container will have a data volume. S3 Clients can connect to any of the 6 minio containers to access the aggregated storage. If you would like to play with distributed XL: https://github.com/minio/minio/tree/distributed

**Comment by krishnasrinivas** [comment]
@sych74 jelastic is super cool! SSH access makes debugging the setup very convenient.

**Comment by sych74** [comment]
@krishnasrinivas, thanks for the quick feedback.
1. how to deploy XL via docker? I guess it's the same docker image and the same latest tag, right?
2. should I just adjust run cmd by replacing this line `/export/volume0 /export/volume1 /export/volume2 /export/volume3 /export/volume4 /export/volume5` to this `server1:/data server2:/data ..... server6:/data` or these arguments are additional? 

Thanks

**Comment by harshavardhana** [comment]
@sych74 you would have to do

```
  6. Start minio server on a 4 node distributed setup. Type the following command on all the 4 nodes.
      $ export MINIO_ACCESS_KEY=minio
      $ export MINIO_SECRET_KEY=miniostorage
      $ minio server 192.168.1.11:/mnt/export/ 192.168.1.12:/mnt/export/ \
          192.168.1.13:/mnt/export/ 192.168.1.14:/mnt/export/
```

A simple docker-compose example shed some light here - https://github.com/minio/minio/blob/distributed/docker-compose.yml 

Distributed branch will be merged to master soon so you would be able to build a docker image to test this properly.

**Comment by harshavardhana** [comment]
Right now we don't have a docker image with distributed branch as soon the master merge completes we will have it as 'docker minio/minio:edge'

**Comment by sych74** [comment]
btw, if I lose one server (hw failure) and add a new one empty, will it sync up all data or stay as is?

**Comment by sych74** [comment]
I build a docker image from distributed branch, here is the error I get now
`+ exec app server --address :80 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T15:36:51Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"
`

my run cmd 

`go-wrapper run server --address :80 minio_1:/export minio_2:/export minio_3:/export minio_4:/export
`

**Comment by harshavardhana** [comment]
> I build a docker image from distributed branch, here is the error I get now
> - exec app server --address :80 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T15:36:51Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1" 
> 
> my run cmd
> 
> go-wrapper run server --address :80 minio_1:/export minio_2:/export minio_3:/export minio_4:/export

Can you run with a different port instead? - i think there is a bug.

**Comment by harshavardhana** [comment]
> btw, if I lose one server (hw failure) and add a new one empty, will it sync up all data or stay as is?

It won't automatically sync up you would have to run "minio control heal" manually, new objects would be created with the new disk btw.

**Comment by sych74** [comment]
Tried different options, still getting the same error
- exec app server --address :81 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T20:03:07Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1" 
- exec app server --address :9000 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T20:47:52Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"  
- exec app server --address 127.0.0.1:9000 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T21:03:28Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"  
- exec app server minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T21:07:11Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"``

**Comment by sych74** [comment]
> It won't automatically sync up you would have to run "minio control heal" manually, new objects would be created with the new disk btw.

Can "minio control heal" be run in parallel with the main minio process? What will happen in theory if we add it to auto start or cron?

**Comment by harshavardhana** [comment]
> Can "minio control heal" be run in parallel with the main minio process? What will happen in theory if we add it to auto start or cron?

You can run `control heal` from any node to any node such as following.  Here 'localhost:9000' can be replaced with your preferred remote. 

```
  1. Heal an object.
     $ minio control heal http://localhost:9000/songs/classical/western/piano.mp3

  2. Heal all objects in a bucket recursively.
     $ minio control heal  http://localhost:9000/songs

  3. Heall all objects with a given prefix recursively.
     $ minio control heal http://localhost:9000/songs/classical/
```

There is no reason to auto-start but it can be scheduled in a cron job.

**Comment by harshavardhana** [comment]
> Tried different options, still getting the same error
> 
> exec app server --address :81 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T20:03:07Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"
> 
> exec app server --address :9000 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T20:47:52Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"
> 
> exec app server --address 127.0.0.1:9000 minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T21:03:28Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"
> 
> exec app server minio_1:/export minio_2:/export minio_3:/export minio_4:/export time="2016-09-13T21:07:11Z" level=fatal msg="Unable to intialize object layer." cause="dial tcp: missing port in address minio_1"``

Can i access this setup through the same account?

**Comment by harshavardhana** [comment]
I see the problem @sych74 it is due to the Dockerfile it is building 'master' for you instead of 'distributed' branch 

Can you check these changes? 

```
diff --git a/Dockerfile b/Dockerfile
index 9dba732..3c44cfd 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -6,7 +6,7 @@ ENV ALLOW_CONTAINER_ROOT=1
 COPY . /go/src/app
 RUN \
        apk add --no-cache git && \
-       go-wrapper download && \
+#      go-wrapper download && \
        go-wrapper install && \
        mkdir -p /export/docker && \
        cp /go/src/app/docs/Docker.md /export/docker/ && \
@@ -14,6 +14,6 @@ RUN \
        apk del git

 EXPOSE 9000
-ENTRYPOINT ["go-wrapper", "run", "server"]
+ENTRYPOINT ["minio", "server"]
 VOLUME ["/export"]
 CMD ["/export"]
```

```
diff --git a/main.go b/main.go
index 532cb44..8eb16fc 100644
--- a/main.go
+++ b/main.go
@@ -14,7 +14,7 @@
  * limitations under the License.
  */

-package main
+package main // import "github.com/minio/minio"

 import minio "github.com/minio/minio/cmd"
```

**Comment by sych74** [comment]
> it is due to the Dockerfile it is building 'master' for you instead of 'distributed' branch

Thanks, my fault. Please let me know when I can use official minio docker image for xl setup test. 

> There is no reason to auto-start but it can be scheduled in a cron job.

I'm thinking about removing the need of manual interactions, if a node fails and then recovers it will be great to have an auto healing...

**Comment by harshavardhana** [comment]
> it is due to the Dockerfile it is building 'master' for you instead of 'distributed' branch
> Thanks, my fault. Please let me know when I can use official minio docker image for xl setup test.
> 
> There is no reason to auto-start but it can be scheduled in a cron job.
> I'm thinking about removing the need of manual interactions, if a node fails and then recovers it will be great to have an auto healing...

@sych74 official `minio/minio:edge` is available with distributed support, we merge the distributed branch yesterday. 

Let us know how it goes @sych74 

> I'm thinking about removing the need of manual interactions, if a node fails and then recovers it will be great to have an auto healing...

This was on purpose to keep the storage simple and predictable, auto healing leads to complexity and unpredictability on a storage system.  We choose N/2 data and parity for this purpose alone that storage shouldn't be doing many things at any given point in time - helps in debugging a system and predictable.  

This sufficiently increases our fault domain leads to higher tolerance towards disk failures and keep the cluster operational for a longer period of time.  As and when you procure fresh disks to replace the failed ones - you can schedule a healing operation. 

Healing comes under maintenance mode which is a scheduled operation, running a maintenance operation in an auto mode all the time complicates a storage system which leads to real issues i.e data loss.

**Comment by sych74** [comment]
I tried edge and got some progress. The automation script has been recently updated https://github.com/jelastic-jps/minio, no more nfs, but it seems like there is a new issue. Only 1 from 4 servers is working http://docker89633-env-5085296.pb-de.jelastic.com/, the rest servers are down, they have similar run configs. I shared this env, could you take a look?

**Comment by harshavardhana** [comment]
> I tried edge and got some progress. The automation script has been recently updated https://github.com/jelastic-jps/minio, no more nfs, but it seems like there is a new issue. Only 1 from 4 servers is working http://docker89633-env-5085296.pb-de.jelastic.com/, the rest servers are down, they have similar run configs. I shared this env, could you take a look?

Will check thanks @sych74

**Comment by harshavardhana** [comment]
> I tried edge and got some progress. The automation script has been recently updated https://github.com/jelastic-jps/minio, no more nfs, but it seems like there is a new issue. Only 1 from 4 servers is working http://docker89633-env-5085296.pb-de.jelastic.com/, the rest servers are down, they have similar run configs. I shared this env, could you take a look?
> Will check thanks @sych74

Fixed the problem - one of the nodes didn't have 'format.json' 

http://docker89633-env-5085296.pb-de.jelastic.com/minio/docker/src/github.com/minio/

Is working and accessible now. 

The procedure for distributed setup is as follows 
- All fresh nodes that need to be part of the cluster should be provided and the same command should be run on all the nodes. 
- Once the fresh disk setup is done a corrupted disk i.e pre-existing data and not having `format.json` or  a fresh disk with no data should be healed using 'minio control heal' 
  (There is a pending bug that needs to be fixed for this currently the console is not reporting that this needs to be run) 

The fix was 'minio control heal http://docker89633-env-5085296.pb-de.jelastic.com/' to heal the disk format. 

There are other problems i observed in UI not displaying total disk usage and server info.  Will close this bug and open relevant bugs. 

Thanks for your patience @sych74 and it is great to see Minio Distributed XL in the dashboard.
